backend/cmd/headlamp.go

```diff
@@ -440,6 +689,148 @@ func StartHeadlampServer(config *HeadlampConfig) {
 	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", config.port), handler)) //nolint:gosec
 }
 
+func GetFreePort() (int, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return 0, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+
+	defer l.Close()
+
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+func getPodsForSvc(svc *corev1.Service, namespace string, k8sClient typev1.CoreV1Interface) (*corev1.PodList, error) {
+	set := labels.Set(svc.Spec.Selector)
+	listOptions := v1.ListOptions{LabelSelector: set.AsSelector().String()}
+	pods, err := k8sClient.Pods(namespace).List(listOptions)
+	return pods, err
+}
+
+// nolint: funlen
+func (c *HeadlampConfig) startPortForward(w http.ResponseWriter, r *http.Request, p PortForwardPayload, token string) error {
+	ports := []string{fmt.Sprintf(p.Port + ":" + p.TargetPort)}
+	ctxtProxy, ok := c.contextProxies[p.Cluster]
+
+	if !ok {
+		return fmt.Errorf("cluster %s not found", p.Cluster)
+	}
+
+	var caData []byte
+
+	var err error
+
+	if caData, err = ctxtProxy.context.cluster.getCAData(); err != nil {
+		return fmt.Errorf("failed to get CA data: %v", err)
+	}
+
+	if token == "" {
+		token = ctxtProxy.context.authInfo.Token
+	}
+	rConf := &rest.Config{
+		Host:        ctxtProxy.context.cluster.config.Server,
+		BearerToken: token,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData:   caData,
+			KeyFile:  ctxtProxy.context.authInfo.ClientKey,
+			CertFile: ctxtProxy.context.authInfo.ClientCertificate,
+			KeyData:  ctxtProxy.context.authInfo.ClientKeyData,
+			CertData: ctxtProxy.context.authInfo.ClientCertificateData,
+		},
+	}
```

Comment:
I'd like us to not have to create a config and use all these + dialer, etc. but we can do that later as we talked on the call.

Reply:
I will create an issue to revisit the portforward implementation